### PR TITLE
add support for assigning roles on Organizations

### PR DESCRIPTION
**Pull request by davepacheco**
This PR adds initial support for assigning roles on Organizations using the interface described in #890 ("option 2").  There's a big block comment I need to update and some missing validation but I'm hopeful that this is close to ready for review.

I'm hopeful that it will be pretty easy to add support for assigning roles on the Fleet, Silos, and Projects because all the underlying pieces here are generic.

### CODE REVIEW

File: common/src/api/external/error.rs [hunk 1]
```diff
@@ -327,8 +327,8 @@ macro_rules! bail_unless {
     };
     ($cond:expr, $($arg:tt)+) => {
         if !$cond {
-            return Err($crate::api::external::Error::internal_error(&format!(
-                $($arg)*)))
+            Err($crate::api::external::Error::internal_error(&format!(
```
Comment (davepacheco) [inline common/src/api/external/error.rs:330]: This kind of strange change makes it possible to use `bail_unless!` in a situation where your function isn't returning `Error` per se, but something that impls `From<Error>`.

File: nexus/src/authz/api_resources.rs [hunk 2]
```diff
@@ -47,38 +47,36 @@ use uuid::Uuid;
 
 /// Describes an authz resource that corresponds to an API resource that has a
 /// corresponding ResourceType and is stored in the database
-pub trait ApiResource: Clone + Send + Sync + 'static {
-    /// If roles can be assigned to this resource, return the type and id of the
-    /// database record describing this resource
+pub trait ApiResource:
```
Comment (davepacheco) [inline nexus/src/authz/api_resources.rs:50]: I found the combination of `ApiResource` and `ApiResourceError` unnecessarily confusing, so I combined them.  This basically meant removing the `Clone + oso::PolarClass` trait bounds here (so that `ApiResource` can be object-safe) and adding them back (actually `oso::ToPolar`) in the few places where we actually need them.

**Comment by davepacheco** [comment]
I think this is close enough to start review.  There are two outstanding issues:
- a test failure apparently due to the use of a generic type that shows up on the OpenAPI spec -- see below
- I need to fix an ascii art comment describing the role_builtin table, now that it has another column

The test failure is:

```
---- integration_tests::commands::test_nexus_openapi stdout ----
writing temp config: /dangerzone/omicron_tmp/test_commands_config.7965.5
thread 'integration_tests::commands::test_nexus_openapi' panicked at 'The type "Policy_for_OrganizationRoles" has a name that is not PascalCase; to rename it add #[serde(rename = "PolicyForOrganizationRoles")]
For more info, see https://github.com/oxidecomputer/openapi-lint#naming

The type "RoleAssignment_for_OrganizationRoles" has a name that is not PascalCase; to rename it add #[serde(rename = "RoleAssignmentForOrganizationRoles")]
For more info, see https://github.com/oxidecomputer/openapi-lint#naming', nexus/tests/integration_tests/commands.rs:124:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    integration_tests::commands::test_nexus_openapi
```

I'm not yet sure how to deal with this.

**Comment by david-crespo** [comment]
I'm confused about that error because I don't see `Policy_for_OrganizationRoles` or `RoleAssignment_for_OrganizationRoles` in the spec. The spec looks like I would expect — one `Policy` and `RoleAssignment` type presumably shared by policies on all resources.

File: openapi/nexus.json [hunk 3]
```diff
@@ -6229,6 +6323,50 @@
...
+            "items": {
+              "$ref": "#/components/schemas/OrganizationRolesRoleAssignment"
+            }
+          }
+        },
+        "required": [
+          "role_assignments"
+        ]
+      },
+      "OrganizationRolesRoleAssignment": {
```
Comment (david-crespo) [inline openapi/nexus.json:6349]: I don't like the thought that every resource will have its own copy of these three types, though if the roles can be different I guess we have no choice?

Comment (davepacheco) [reply openapi/nexus.json:6349]: Yeah, I think this is a tradeoff for having the allowed roles be expressed in the API.  If there's another way to do it without that tradeoff that'd be great!

Comment (david-crespo) [reply openapi/nexus.json:6349]: Will all such types be the same except for their names?

**Review by plotnick — APPROVED** [review]
This looks great; thanks for tackling this very thorny issue! One abstract thought that I had reading it was that a `Policy` can be more general than just a list of role assignments, but I don't think there's anything in here that limits us to just that if we need to expand the notion later. I also worry slightly about keeping the several enums for identity/actor type in sync, but don't have a better alternative at the moment.

File: openapi/nexus.json (hunk 3 above)
Comment (david-crespo) [reply openapi/nexus.json:6349]: If they will, then one approach would be to `impl From<RolesPolicy> for OrganizationRolesPolicy` on all the individual types. That would give them all the same type in the OpenAPI spec.

Comment (david-crespo) [reply openapi/nexus.json:6349]: But I see what we're buying with the separate types: a generated client will already know at the type level what roles are valid for a given resource, which is pretty nifty.

```ts
type OrganizationRoles = 'Admin' | 'Collaborator'

type OrganizationRolesPolicy = {
  roleAssignments: OrganizationRolesRoleAssignment[]
}

type OrganizationRolesRoleAssignment = {
  identityId: string
  identityType: IdentityType
  roleName: OrganizationRoles
}
```

I guess I'm basically lamenting that OpenAPI doesn't let us express this relationship (as far as I know):

```ts
type RolesPolicy<Roles> = {
  roleAssignments: RoleAssignment<Roles>[]
}

type RoleAssignment<Roles> = {
  identityId: string
  identityType: IdentityType
  roleName: Roles
}

type OrganizationRoles = 'Admin' | 'Collaborator'
// use RoleAssignment<OrganizationRoles> wherever

type ProjectRoles = 'Admin' | 'Collaborator'
// use RoleAssignment<ProjectRoles> wherever
```

**Comment by davepacheco** [comment]
> I'm confused about that error because I don't see `Policy_for_OrganizationRoles` or `RoleAssignment_for_OrganizationRoles` in the spec. The spec looks like I would expect — one `Policy` and `RoleAssignment` type presumably shared by policies on all resources.

The reason for this is that the test that updates the OpenAPI spec checks lint *first*.  So the copy of the spec in the repo for that commit isn't what was being linted.  To show the spec that was generated without the serde rename, I generated this commit:
https://github.com/oxidecomputer/omicron/commit/20695844b154f80dd9d4077cb0411a0d5279a3f5

In that one, the test checks lint _after_ updating the local copy, so that spec is what's being linted.

File: nexus/src/external_api/shared.rs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+            panic!(
+                "expected internal error for database parse failure, \
+                found {:?}",
+                error
+            );
+        }
+
+        let success = <shared::RoleAssignment<DummyRoles>>::try_from(ok_input)
+            .expect("parsing valid role assignment from database");
+        println!("success: {:?}", success);
```
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/src/external_api/shared.rs:216]: seems like useful to see if the test fails?

[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/external_api/shared.rs:216]: I actually kept this on purpose so that if you were debugging this test you could see the output.  It's hidden unless you use `--nocapture` or the test fails.

File: nexus/src/authz/api_resources.rs [hunk 5]
```diff
@@ -321,6 +331,24 @@ authz_resource! {
...
+    DeserializeFromStr,
+    Display,
+    Eq,
+    FromStr,
+    PartialEq,
+    SerializeDisplay,
+    JsonSchema,
+)]
+#[display(style = "kebab-case")]
+pub enum OrganizationRoles {
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/authz/api_resources.rs:347]: Yeah, that's fair.  I considered making this an argument to the `authz` macro instead.  I'd basically change `roles_allowed = true` to `allowed_roles = <an enum type>`.

Alternatively, I could remove the reference to `{Foo}Roles` from the macro altogether -- it'd be pretty reasonable to just impl the `ApiResourceWithRolesType` trait explicitly for the few types where we want it.

File: nexus/src/authz/roles.rs [hunk 6]
```diff
@@ -132,19 +134,20 @@ pub async fn load_roles_for_resource(
     roleset: &mut RoleSet,
 ) -> Result<(), Error> {
     // If the user is authenticated ...
-    if let Some(actor_id) = authn.actor() {
+    if let Some(actor) = authn.actor() {
         // ... then fetch all the roles for this user that are associated with
         // this resource.
         trace!(opctx.log, "loading roles";
-            "actor_id" => actor_id.id.to_string(),
+            "actor" => ?actor,
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/authz/roles.rs:141]: Good question.  The reason I changed this is to capture other information about the actor (like the silo user id).  This might differ for different variants of `Actor`.  How about if I `impl Debug for Actor` explicitly with a comment about not including anything sensitive?  That way we'll hopefully catch ourselves from leaking something private, but people can still freely use `Debug` output for objects like this?

File: nexus/src/authz/api_resources.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/authz/api_resources.rs:351]: I did the latter in a447ae61.  I like this better -- it also means I didn't have to generate unused enums for Silos and Projects (those will come in a follow-on PR, but I previously had them in this PR because the macro did this).  Thanks for raising this.

File: nexus/src/authz/roles.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/authz/roles.rs:141]: Done in 8c625082a705523338886c1d08eed7f63d4fcbcb.

**Comment by davepacheco** [comment]
> One abstract thought that I had reading it was that a `Policy` can be more general than just a list of role assignments, but I don't think there's anything in here that limits us to just that if we need to expand the notion later.

That's right.  For the MVP it's just that, but the intent is to be able to expand it later if we want.

> I also worry slightly about keeping the several enums for identity/actor type in sync, but don't have a better alternative at the moment.

I share that worry, but it seems to be the pattern we're using for better or worse.  I copied the pattern from nexus/src/db/model.rs, where we do something similar for DatasetKind and InstanceState.  At least Rust will tell us if either grows a variant that isn't handled in the other.

File: openapi/nexus.json (hunk 3 above)
Comment (davepacheco) [reply openapi/nexus.json:6349]: Yeah, it would be neat if OpenAPI let you do that.

To summarize: my primary intent here was to have the Nexus type reflect the set of allowed role names.  That way, this validation is handled like all other validation, while parsing the input in the first place, and the other code paths can know from the type that the role names have been validated.

I went the logical next step of using this type directly in the HTTP handler functions (both input and output) so that the OpenAPI spec reflects the allowed values, especially since we _do_ validate that so we may as well document what the accepted values are in the spec.  As you noted, this means the generated clients can know exactly which values are allowed, which is cool.  We don't _have_ to do this part though.  We could still accept and return strings instead of enums.  And it's possible that in the future we'll want to do that because if we support custom roles, then there won't be an enum of allowed values.  But I'd rather not spend much time now worrying about that for the usual reasons: it's hard to know we're actually solving the future problem no matter what we do now, and we can always evolve this API compatibly anyway.

Comment (ahl) [reply openapi/nexus.json:6353]: @david-crespo is your lament, in essence, that OpenAPI / JSON Schema doesn't permit the expression of generics types / types with parameters?

**Review by ahl — COMMENTED** [review]
for posterity now since I realize this is already merged...

File: nexus/src/authz/api_resources.rs [hunk 7]
```diff
@@ -321,6 +331,24 @@ authz_resource! {
...
+    Debug,
+    DeserializeFromStr,
+    Display,
+    Eq,
+    FromStr,
+    PartialEq,
+    SerializeDisplay,
+    JsonSchema,
+)]
+#[display(style = "kebab-case")]
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/authz/api_resources.rs:350]: Agreed.  I don't think I care between snake_case and kebab-case.  I'm sure I copied this from somewhere else thinking it was our standard.  Yes, it'd be great to lint it!

File: nexus/src/db/datastore.rs [hunk 8]
```diff
@@ -2845,6 +2848,115 @@ impl DataStore {
...
+    /// Removes all existing role assignments on `authz_resource` and adds those
+    /// specified by `new_assignments`
+    ///
+    /// The expectation is that the caller will have just fetched the role
+    /// assignments, modified them, and is giving us the complete new list.
+    ///
+    /// This function is generic over all resources that can accept roles (e.g.,
+    /// Fleet, Silo, Organization, etc.).
+    // TODO-correctness As with the rest of the API, we're lacking an ability
+    // for an ETag precondition check here.
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/db/datastore.rs:2891]: I suspect there's a lot of work here but I don't know.  I feel like several times over the last year I've thought "ugh, <thing I'm working on> is gonna get complicated when we implement etags".  But thinking out loud:

- We need to decide how to generate the etag for every kind of resource.  Is it random (and updated with every modification) or content-based?  If content-based, is it stored in the database or calculated when needed?  If calculated when needed, it'd be nice if we could do it in SQL so that you can calculate it and check it within a SQL query (see below).  What fields go into it?  How do we specify that on a per-resource basis?  How do we keep it up to date if you're doing something like `update ... set ... where <condition that matches 10 rows>`?
- We need to plumb it through on "list" and "get" for everything.  Gotta figure out where to store it...if it's stored in
the database, great.  If not, how do we pass it around?  Maybe it goes into the `authz` object?  That might sound weird but the `authz` object basically represents a handle to a specific object, regardless of how it was looked up or what it's called now.  It makes sense that would also be locked to a specific revision of that object.
- We need to modify all the "update" datastore functions so that they accept an optional etag and do the conditional update.  In an ideal world, we'd issue a SQL "update" with the etag condition in it (rather than read-check-write).
- We need to plumb that optional etag argument out to all the PUT entrypoints.

That doesn't sound so bad.  We might need better transaction support (see #973) -- I'm not sure.

File: nexus/src/external_api/shared.rs [hunk 9]
```diff
@@ -0,0 +1,221 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Types that are used as both views and params
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/external_api/shared.rs:5]: I agree that the current split is confusing.  I understand the current guidance to be:

* if it's an input param, it goes in "params"
* if it's an output, it goes in "views"
* if it's both, it goes in "shared"
* there's also a bunch of stuff in "common" that hasn't been put into one of those three places because it's hard

This last point makes it kind of a mess, and it's also possible that this last part is hard because there are other constraints involved that this guidance doesn't accommodate.  Progenitor changes a bunch of this too -- I think part of the original reason for "common" was for stuff that was shared between clients and servers of a given service.

On your specific example:

> We could, for example, have groups of endpoints and types live together.

We could.  Any way I can think to group them sucks a little bit -- including the current approach of putting them in one big file.  (If we group them by resource, one might be sad that the various "create" endpoints (which are very parallel to each other and use some of the same types, like IdentityMetadataCreate) don't live near each other.  Also, it's kind of nice for some small groups of resources to live together (e.g., built-in users and roles, which might go under an "IAM" bucket).  So maybe this is by tag after all?)

Basically, I don't have a strong idea of how these should be better organized.

File: nexus/src/external_api/shared.rs [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+// customers to develop a dependency on a huge number of role assignments.  That
+// maximizes our flexibility in the future.
+//
+// TODO This should be runtime-configurable.  But it doesn't belong in the Nexus
+// configuration file, since it's a constraint on database objects more than it
+// is Nexus.  We should have some kinds of config that lives in the database.
+pub const MAX_ROLE_ASSIGNMENTS_PER_RESOURCE: usize = 64;
+
+/// Client view of a [`Policy`], which describes how this resource may be
+/// accessed
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/external_api/shared.rs:33]: 💯 agree -- and I think that'll be true for all the stuff in `views`.  (I copied the form there for these types.)  I think we'll want to make an edit pass that documents all these things properly.

File: openapi/nexus.json (hunk 3 above)
Comment (david-crespo) [reply openapi/nexus.json:6353]: Yes — having identical `*RolePolicy` and `*RoleAssignment` types for every resource feels silly. But it seems there's no way around it.
